### ISSUE TITLE
Weird loop 

### ISSUE THREAD

**Issue by angelhdzmultimedia** [caoxiemeihao/nuxt-electron#74]
I added nuxt-electron to my fully functional nuxt app, the window opens, Nuxt loads, then it starts a weird loop and tons of errors in the terminal and browser console... so I created a fresh Nuxt app and added nuxt-electron and the same. Downgraded electron from v30 to v28, same. 😭

**Comment by caoxiemeihao** [comment]
I ran into the same issue and am trying to find the cause of the BUG in nuxt's source code.
At present, I found that the nuxt startup port is 3000 by default, but in the hooks of `nuxt-electron`, I always get a random port, which will cause Electron to fall into an infinite loop after startup.

<img width="656" alt="image" src="https://github.com/caoxiemeihao/nuxt-electron/assets/26263658/8a40ff5f-d44e-483d-ac85-3d0796679586">

- `nuxt-electron` hooks
https://github.com/caoxiemeihao/nuxt-electron/blob/9661c57047bee1609a036f9ce2cd8230b532d592/src/index.ts#L110

- `nuxt` source code
https://github.com/unjs/listhen/blob/v1.7.2/src/listen.ts#L102

---

I'm stuck on this, maybe someone can give me some help? Thanks in advance. @danielroe 🙏

**Comment by caoxiemeihao** [comment]
Eventually I found out that this was because `nuxt`’s [Performance](https://github.com/nuxt/cli/blob/v3.11.1/CHANGELOG.md#-performance) mode would fork the main process once, causing `nuxt-electron`’s hooks to listen to the wrong(random) `server.port`.

**I think this is an issue that needs improvement.**

Fork main process https://github.com/nuxt/cli/blob/v3.11.1/src/commands/dev.ts#L68-L73
<img width="603" alt="image" src="https://github.com/caoxiemeihao/nuxt-electron/assets/26263658/292d21ce-8392-4863-ac28-d1e7c4e050b8">

`hooks.listen` only runs in the child process after the fork, causing `server.port` to become a random number. https://github.com/nuxt/cli/blob/v3.11.1/src/utils/dev.ts#L262-L266
<img width="390" alt="image" src="https://github.com/caoxiemeihao/nuxt-electron/assets/26263658/030b8919-7657-47e0-af96-04455d8c752b">

---

Currently, we can temporarily avoid this issue in the following ways.

```json
{
  "scripts": {
-   "dev": "nuxi dev",
+   "dev": "nuxi dev --no-fork",
    "build": "nuxi build --prerender && electron-builder"
  }
}
```

**Comment by angelhdzmultimedia** [comment]
> Eventually I found out that this was because `nuxt`’s [Performance](https://github.com/nuxt/cli/blob/v3.11.1/CHANGELOG.md#-performance) mode would fork the main process once, causing `nuxt-electron`’s hooks to listen to the wrong(random) `server.port`.
> 
> **I think this is an issue that needs improvement.**
> 
> Fork main process https://github.com/nuxt/cli/blob/v3.11.1/src/commands/dev.ts#L68-L73 <img alt="image" width="603" src="https://private-user-images.githubusercontent.com/26263658/323280079-292d21ce-8392-4863-ac28-d1e7c4e050b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzODgxMzUsIm5iZiI6MTcxMzM4NzgzNSwicGF0aCI6Ii8yNjI2MzY1OC8zMjMyODAwNzktMjkyZDIxY2UtODM5Mi00ODYzLWFjMjgtZDFlN2M0ZTA1MGI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDIxMDM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmOWI0MzgzNTRjNjY4MTYzY2U5MTU5NGMwNDlmYTc4M2YxYjkxZjIzZjk2ZmQ4ZGM1NzMwNWQ2OGE2YTI4NmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.w4m6rti2ilvlx54_MKtYIuTcX69AQttrSfxArtJJmX4">
> 
> `hooks.listen` only runs in the child process after the fork, causing `server.port` to become a random number. https://github.com/nuxt/cli/blob/v3.11.1/src/utils/dev.ts#L262-L266 <img alt="image" width="390" src="https://private-user-images.githubusercontent.com/26263658/323282206-030b8919-7657-47e0-af96-04455d8c752b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzODgxMzUsIm5iZiI6MTcxMzM4NzgzNSwicGF0aCI6Ii8yNjI2MzY1OC8zMjMyODIyMDYtMDMwYjg5MTktNzY1Ny00N2UwLWFmOTYtMDQ0NTVkOGM3NTJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDIxMDM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0OTI1ZGE1YjI2ZWU4ZTk5NmIzOGI2ZDY1NzNhOTBhMjZjMzhmZjE3NTQ1ZmJlZTA0YzgzZmY0MzZiMGU5OTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xZLSplzJa3DAJgYSzDGFBB2MJd5YiCEe0njFivfZL_U">
> 
> Currently, we can temporarily avoid this issue in the following ways.
> 
> ```diff
> {
>   "scripts": {
> -   "dev": "nuxi dev",
> +   "dev": "nuxi dev --no-fork",
>     "build": "nuxi build --prerender && electron-builder"
>   }
> }
> ```

Thank you so much. Will try in a few hours.
Hope you have a nice rest of the day. Lots of health, peace, and success.

**Comment by angelhdzmultimedia** [comment]
### Update
Modified the `dev` script as you mentioned. Error continues. But found something interesting... if I run `bun --bun run dev` (which runs the script with Bun runtime instead of Node), it works right away and fast!!!

Node:

![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10191085/0a5ecdd0-c25a-4f35-b5fa-72122baa08ab)
```
[25552:0417/174211.789:ERROR:CONSOLE(1)] "Request Autofill.enable failed. {"code":-32601,"message":"'Autofill.enable' wasn't found"}", source: devtools://devtools/bundled/core/protocol_client/protocol_client.js (1)
```
Bun:

![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10191085/8427da37-92b0-4c63-a0c6-5ba5c065aa0d)

**Comment by jbhaywood** [comment]
For anyone else that comes here for this issue, @caoxiemeihao's fix worked for me. I'm not using Bun, fwiw.

**Comment by angelhdzmultimedia** [comment]
> For anyone else that comes here for this issue, @caoxiemeihao's fix worked for me. I'm not using Bun, fwiw.

I keep having the issue, switched to pnpm. Deleted `node_modules`. Ran `pnpm install` and `pnpm run dev` which points to the `dev` script with the `--no-fork` boolean flag. 😔

**Comment by angelhdzmultimedia** [comment]
Hmmm... the loop is gone... new error: 
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10191085/2d37b0c3-1fab-4380-ad52-6d089d9ead8c)

**Comment by caoxiemeihao** [comment]
You can try to use the latest **quick-start**, I downgraded `nuxt` to `3.6.5` and everything works fine.

**Comment by danielroe** [comment]
The issue here is not Nuxt but a change in behavior in `nuxi` itself, which is now [versioned separately](https://github.com/nuxt/cli). You can likely downgrade your version of `nuxi` - but note there are some issues in that repo that explain what's happening and likely give some tips on how to work around it.

**Comment by angelhdzmultimedia** [comment]
> @angelhdzmultimedia You can try to use the latest **quick-start**, I downgraded `nuxt` to `3.6.5` and everything works fine.

Oooh downgrading to nuxt 3.6.5 did it! Thanks! 🤯🙏
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10191085/8c856506-27c5-4fed-af34-c4bc350b0aec)

**Comment by Patrity** [comment]
Thanks for developing this great module @caoxiemeihao I was reliant on it for a couple of my apps, but currently can not use it without downgrading multiple packages which are needed.
The `--no-fork` flag did not end up working for me, but I'm commenting to stay in the loop on hopefully any future updates/fixes.

**Comment by angelhdzmultimedia** [comment]
> Thanks for developing this great module @caoxiemeihao I was reliant on it for a couple of my apps, but currently can not use it without downgrading multiple packages which are needed. The `--no-fork` flag did not end up working for me, but I'm commenting to stay in the loop on hopefully any future updates/fixes.

You also need to downgrade nuxt package to 3.6.5

**Comment by Patrity** [comment]
> > Thanks for developing this great module @caoxiemeihao I was reliant on it for a couple of my apps, but currently can not use it without downgrading multiple packages which are needed. The `--no-fork` flag did not end up working for me, but I'm commenting to stay in the loop on hopefully any future updates/fixes.
> 
> You also need to downgrade nuxt package to 3.6.5

As I mentioned, I am using features in both Nuxt and other libraries that depend on newer versions. Thank you for the reply though.

**Comment by Hauskrampf** [comment]
I also encountered this loop problem, tried @angelhdzmultimedia's bun solution with @caoxiemeihao extra flag, which worked **once**, although the DevTools where broken.
Any builds since then just spits out, in the electron window:

> JavaScript execution terminated.
> 
> Stack trace:
> ```js
> 
>   λ()
>   λ()
>   λ()
>   λ()
>   λ()
>   λ()
>   λ()
>   λ()
> ```
> Info:
>   > bun v1.1.4
>   > wsl #1 SMP Thu Oct 5 21:02:42 UTC 2023 (x64)
>   > User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) nuxt-app/0.0 Chrome/124.0.6367.60 Electron/30.0.1 Safari/537.36
>   > Pathname: /
> 

and the console prints:
```bash
error: JavaScript execution terminated.
GET - / failed
```

I then tried to remove the `--no-fork` flag, which now successfully builds sometimes?
Other times i get:
```bash
519 |       return this.emit(
520 |         "error",
521 |         new Error("Must provide a proper URL as target")
522 |       );
523 |     }
524 |     let _resolve;
             ^
error: Must provide a proper URL as target
      at /home/[etc pp]/node_modules/httpxy/dist/index.mjs:524:8
      at /home/[etc pp]/node_modules/nuxi/dist/chunks/dev2.mjs:6902:29
      at /home/[etc pp]/node_modules/nuxi/dist/chunks/dev2.mjs:6891:21
      at emit (node:events:158:48)
      at fetch (node:http:483:12)
error: script "dev" exited with code 1
```

The sheer randomness  of this bug baffles me....

TLDR: If you don't want to downgrade nuxt, try `bun --bun run dev`, it may work ¯\_(ツ)_/¯

**Comment by angelhdzmultimedia** [comment]
> I also encountered this loop problem, tried @angelhdzmultimedia's bun solution with @caoxiemeihao extra flag, which worked **once**, although the DevTools where broken. Any builds since then just spits out, in the electron window:
> 
> > JavaScript execution terminated.
> > Stack trace:
> > ```js
> >   λ()
> >   λ()
> >   λ()
> >   λ()
> >   λ()
> >   λ()
> >   λ()
> >   λ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Info:
> > > bun v1.1.4
> > > wsl #1 SMP Thu Oct 5 21:02:42 UTC 2023 (x64)
> > > User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) nuxt-app/0.0 Chrome/124.0.6367.60 Electron/30.0.1 Safari/537.36
> > > Pathname: /
> 
> and the console prints:
> 
> ```shell
> error: JavaScript execution terminated.
> GET - / failed
> ```
> 
> I then tried to remove the `--no-fork` flag, which now successfully builds sometimes? Other times i get:
> 
> ```shell
> 519 |       return this.emit(
> 520 |         "error",
> 521 |         new Error("Must provide a proper URL as target")
> 522 |       );
> 523 |     }
> 524 |     let _resolve;
>              ^
> error: Must provide a proper URL as target
>       at /home/[etc pp]/node_modules/httpxy/dist/index.mjs:524:8
>       at /home/[etc pp]/node_modules/nuxi/dist/chunks/dev2.mjs:6902:29
>       at /home/[etc pp]/node_modules/nuxi/dist/chunks/dev2.mjs:6891:21
>       at emit (node:events:158:48)
>       at fetch (node:http:483:12)
> error: script "dev" exited with code 1
> ```
> 
> and the DevTools are still broken.
> 
> The sheer randomness of this bug baffles me....
> 
> TLDR: If you don't want to downgrade nuxt, try `bun --bun run dev`, it may work ¯_(ツ)_/¯

The reason it works with --bun flag is it's running with Bun not Node. The bug seems to be related to higher versions of Nuxt + Node.  Hope this is fixed soon. I stopped working on my electron app due to this.

**Comment by rvalledorjr** [comment]
Encountered the same issue on a new project and able to resolve it right now by my updating `nuxt.config.ts` to:

```ts
export default defineNuxtConfig({
  devtools: { enabled: true },
  modules: [
    (options, nuxt) => {
      if (!nuxt.options.dev) {
        nuxt.options.nitro.runtimeConfig ??= {}
        nuxt.options.nitro.runtimeConfig.app ??= {}
        nuxt.options.nitro.runtimeConfig.app.baseURL = './'
      }
    },
    'nuxt-electron',
  ],
  electron: {
    disableDefaultOptions: true,
  },
  ssr: false,
  router: {
    options: {
      hashMode: true
    }
  },
})
```

* Had to do `disableDefaultOptions: true` as this [line](https://github.com/caoxiemeihao/nuxt-electron/blob/d7cf10ddaf8e050564d46b0a906cc7a468573e60/src/index.ts#L172) seems to cause the reload loop.
* to retain the other configs, I put them back again and added an inline module to retain those that need only be applied on build mode.

**Comment by kogratte** [comment]
Good evening all!

I ran into the same issue today, but --no-fork didn't do the job.
I've been able to workaround the issue by using distinct launch.

```Makefile

devserver: ## Run dev server
	yarn dev

develectron: ## Run dev electron
	curl --head -X GET --retry 5 --retry-connrefused --retry-delay 1 http://localhost:3000
	yarn run dev:electron

buildelectron: ## Build app for electron
	npx nuxi build

dev: install ## Run dev server
	test -s ./dist-electron/main.js || { make buildelectron; }
	$(MAKE) -j2 devserver develectron
.PHONY: dev
```

```package.json
{
   [...],
   scripts: {
       "dev": "DEV_MODE=1 nuxt dev",
        "dev:electron": "VITE_DEV_SERVER_URL=http://localhost:3000 electron ."
  }
  [...]
}
```

And finally...

```nuxt.config.ts
export default defineNuxtConfig({
  modules: [
    process.env.DEV_MODE ? undefined : 'nuxt-electron'
  ],
```

The`make dev` ensure that electron target has been built at least once (to avoid a missing dist-electron), otherwise build it.

Once done, it runs both a devserver, and an electron window.
The electron window try to wait for the vite server to be ready prior opening it.

You can close the whole thing using a classic `cmd+c`.

**Comment by kingyue737** [comment]
@rvalledorjr  provides a good solution. Here we can make it simpler. Remove the inline module and set the production-only option `cdnURL` can achieve the same effect:
```diff
// Only relevant configs are shown
export default defineNuxtConfig({
  modules: [
-   (options, nuxt) => {
-     if (!nuxt.options.dev) {
-       nuxt.options.nitro.runtimeConfig ??= {}
-       nuxt.options.nitro.runtimeConfig.app ??= {}
-       nuxt.options.nitro.runtimeConfig.app.baseURL = './'
-     }
-   },
    'nuxt-electron',
  ],
  electron: {
    disableDefaultOptions: true,
  },
+ app:{ cdnURL: './'},
  ssr: false,
  router: {
    options: {
      hashMode: true
    }
  },
})
```

**Comment by necro-wbj** [comment]
I resolved the issue and wanted to share my solution here.

I initially noticed that the problem was caused by inconsistent ports, which led to repeated restarts.

I found that by applying the following configuration, the restarts could be prevented:
```javascript
electron: {
    ...
    disableDefaultOptions: true
},
```
I looked into what the disableDefaultOptions parameter does and found that it modifies the following:
```
{
    ssr: false,
    app: {
        baseURL: './',
        buildAssetsDir: '/',
    },
    runtimeConfig: {
        app: {
            baseURL: './',
            buildAssetsDir: '/',
        },
    },
    nitro: {
        runtimeConfig: {
            app: {
                baseURL: './',
            }
        }
    },
}
```
I tried to figure out which parameter affected the outcome, and I eventually discovered that the app setting differed from the default value defined on the [Nuxt documentation](https://nuxt.com/docs/api/nuxt-config#baseurl), leading to this result:
```
app: {
    baseURL: '/'
}
```
But the issue wasn’t entirely resolved. I encountered problems with image loading, and my final configuration was:
```
ssr: false,
app: {
    baseURL: '/',
    buildAssetsDir: '/',  // This differs from the Nuxt documentation but is closer to what nuxt-electron expects.
},
```
I also found a related issue on the Nuxt issue tracker where someone reported problems using ./:
[issues#28624](https://github.com/nuxt/nuxt/issues/28624)

**Comment by kingyue737** [comment]
@necro-wbj You can't set `baseURL: '/'` globally. It doesn't work in production as electron will load the entry `index.html` from file system. The correct way is mentioned in https://github.com/caoxiemeihao/nuxt-electron/issues/74#issuecomment-2295093963

**Comment by necro-wbj** [comment]
> @necro-wbj You should not set `baseURL: '/'` globally. It doesn't work in production as electron will load the entry `index.html` from file system. The correct way is mentioned in [#74 (comment)](https://github.com/caoxiemeihao/nuxt-electron/issues/74#issuecomment-2117567168) and [#74 (comment)](https://github.com/caoxiemeihao/nuxt-electron/issues/74#issuecomment-2295093963)

Thank you for the reminder. I just resolved the issue and haven't had the chance to test it in the production environment.

**Comment by michaeljfazio** [comment]
> @rvalledorjr provides a good solution. Here we can make it simpler. Remove the inline module and set the production-only option `cdnURL` can achieve the same effect:
> 
> ```diff
> // Only relevant configs are shown
> export default defineNuxtConfig({
>   modules: [
> -   (options, nuxt) => {
> -     if (!nuxt.options.dev) {
> -       nuxt.options.nitro.runtimeConfig ??= {}
> -       nuxt.options.nitro.runtimeConfig.app ??= {}
> -       nuxt.options.nitro.runtimeConfig.app.baseURL = './'
> -     }
> -   },
>     'nuxt-electron',
>   ],
>   electron: {
>     disableDefaultOptions: true,
>   },
> + app:{ cdnURL: './'},
>   ssr: false,
>   router: {
>     options: {
>       hashMode: true
>     }
>   },
> })
> ```

This fixed all issues for me.
